### PR TITLE
[CIS-108] Improve WebSocket tests

### PR DESCRIPTION
**Pull request by VojtaStavik**
### In this PR: 

- New `Timer` wrapper type for timers. This allows us to inject custom timers when running tests.

- It's now possible to inject `Timer.type` to `WebSocket`.

- Introduced `VirtualTime` for running timers in tests without actually waiting. The implementation of the `run(_:)` method is a little bit tricky. Please see `VirtualTimeTests` if you want to understand why it's done like this. Also, happy to see your suggestions and improvements.

- Refactored existing `WebSocket` tests to use virtual time. I didn't refactor the tests for `Event` callbacks because we'll change this part soon and we'll add tests as a part of the refactoring.

#no_changelog

### CODE REVIEW

File: Tests/Client/Unit Tests/WebSocket/WebSocketProviderMock.swift [hunk 1]
```diff
@@ -2,125 +2,61 @@
...
+    }
     
-    init(request: URLRequest, callbackQueue: DispatchQueue) {
+    required init(request: URLRequest, callbackQueue: DispatchQueue) {
         self.request = request
         self.callbackQueue = callbackQueue
     }
     
+    /// How many times was `connect()` called
+    var connectCalledCount = 0
```
Comment (b-onc) [inline Tests/Client/Unit Tests/WebSocket/WebSocketProviderMock.swift:30]: Can we group variables and functions? Comment makes it obvious anyway

File: Tests/Client/Unit Tests/WebSocket/WebSocketProviderMock.swift [hunk 2]
```diff
@@ -2,125 +2,61 @@
...
-        isConnected = false
-        connectionId = nil
-        stopTimer()
-        let error = WebSocketProviderError(reason: "stop", code: 1000, providerType: Self.self, providerError: nil)
-        delegate?.websocketDidDisconnect(error: error)
+}
+
+// MARK: - Functions to simulate behavior
+
+extension WebSocketProviderMock {
```
Comment (b-onc) [inline Tests/Client/Unit Tests/WebSocket/WebSocketProviderMock.swift:48]: Can we just use the mark inside main class and not an extension?

File: Tests/Client/Unit Tests/WebSocket/WebSocketTests.swift [hunk 3]
```diff
@@ -11,124 +10,140 @@ import XCTest
...
     
-    func test_webSocket_reconnection() {
-        // Expect 3 connected events. 2 from reconnection.
-        let connectedExpectation = expectation(description: "WebSocket connected twice")
-        connectedExpectation.expectedFulfillmentCount = 3
-        let reconnectingExpectation = expectation(description: "WebSocket reconnecting")
-        reconnectingExpectation.expectedFulfillmentCount = 2
+    func test_reconnectionFlow_withoutStopError() {
+        // Setup (connect)
+        test_connectionFlow()
```
Comment (b-onc) [inline Tests/Client/Unit Tests/WebSocket/WebSocketTests.swift:57]: We shouldn't call other tests within a test case, just `connect` and `simulateconnect` would be fine

File: Tests/Client/Unit Tests/WebSocket/WebSocketTests.swift [hunk 4]
```diff
@@ -11,124 +10,140 @@ import XCTest
...
+        // Assert
+        AssertAsync {
+            Assert.staysTrue(self.socketProvider.connectCalledCount == originalConnectCalledCount)
         }
+    }
+
+    func test_pingIsSentPeriodically() {
+        // Setup (connect)
+        test_connectionFlow()
+        assert(WebSocket.pingTimeInterval == 25)
```
Comment (b-onc) [inline Tests/Client/Unit Tests/WebSocket/WebSocketTests.swift:108]: I don't think this makes sense here. We shouldn't assume ping interval will be 25, and write the test with accessing this value (calculated)

File: Tests/Client/Unit Tests/WebSocket/WebSocketProviderMock.swift (hunk 1 above)
Comment (VojtaStavik) [reply Tests/Client/Unit Tests/WebSocket/WebSocketProviderMock.swift:30]: What do you mean?

Comment (VojtaStavik) [reply Tests/Client/Unit Tests/WebSocket/WebSocketProviderMock.swift:30]: ping @b-onc

Comment (b-onc) [reply Tests/Client/Unit Tests/WebSocket/WebSocketProviderMock.swift:30]: Putting variables on top, functions after them, and not mixing the order

File: Sources/Client/WebSocket/WebSocket.swift [hunk 5]
```diff
@@ -95,13 +101,13 @@ extension WebSocket {
...
+        let maxDelay: TimeInterval = min(500 + consecutiveFailures * 2000, 25000)
+        let minDelay: TimeInterval = min(max(250, (consecutiveFailures - 1) * 2000), 25000)
         consecutiveFailures += 1
-        let delay = minDelay + TimeInterval.random(in: 0...(maxDelay - minDelay))
-        logger?.log("⏳ Reconnect in \(delay) sec")
+        let delay = TimeInterval.random(in: minDelay...maxDelay)
+        logger?.log("⏳ Reconnect in \(delay / 1000) sec")
         
-        provider.callbackQueue.asyncAfter(deadline: .now() + delay) { [weak self] in
+        Timer.schedule(timeInterval: .milliseconds(Int(delay)), queue: provider.callbackQueue) { [weak self] in
```
[the remark this answers is not in the document]
Comment (VojtaStavik) [reply Sources/Client/WebSocket/WebSocket.swift:109]: `delay` is `TimeInterval` but `Timer` access `DispatchTimeInterval`. `DispatchTimeInterval.miliseconds(_:)` accepts only `Int` 🤷‍♂️

[the remark this answers is not in the document]
Comment (b-onc) [reply Sources/Client/WebSocket/WebSocket.swift:109]: We have `TimeInterval.init(dispatchTimeInterval: DispatchTimeInterval)` so why not the other way? `DispatchTimeInterval.init(timeInterval: TimeInterval)`

[the remark this answers is not in the document]
Comment (VojtaStavik) [reply Sources/Client/WebSocket/WebSocket.swift:109]: I don't know man. Seems unnecessary to me to add a completely new initializer just to get rid of `Int(___)` . I think it's fine as is tbh.

[the remark this answers is not in the document]
Comment (b-onc) [reply Sources/Client/WebSocket/WebSocket.swift:109]: Not something I'd push ¯\_(ツ)_/¯

File: Tests/Virtual Time/VirtualTimers.swift [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+extension VirtualTime.Timer: TimerControl {
+    func resume() {
+        isActive = true
+    }
+    
+    func suspend() {
+        isActive = false
+    }
+}
+
```
[the remark this answers is not in the document]
Comment (VojtaStavik) [reply Tests/Virtual Time/VirtualTimers.swift:39]: I think there's nothing wrong about setting `isActive` directly in the tests. No need to complicate things. This is not public api.

[the remark this answers is not in the document]
Comment (b-onc) [reply Tests/Virtual Time/VirtualTimers.swift:39]: Why would we use internal API in tests? It's safer to be close to real-life usage scenario, in which we'll never access `isActive` directly. I also disagree it's complicating, on the contrary I think accessing `isActive` is more complicated than just `timer.suspend()` or `resume()`

[the remark this answers is not in the document]
Comment (VojtaStavik) [reply Tests/Virtual Time/VirtualTimers.swift:39]: Bahadir, we're talking about `VirtualTimer` which is used solely in tests, never in production. If mutating a variable directly bothers you for some reason, I'll add the functions you want 🤷‍♂️

[the remark this answers is not in the document]
Comment (b-onc) [reply Tests/Virtual Time/VirtualTimers.swift:39]: That's a fair point, but then why do we have the extensions? If we have functions to mutate these variables, we should only use them.
Even if it's testing something only used for testing, API could be consistent.
I accept your point though, it's up to you, we can keep it.

[the remark this answers is not in the document]
Comment (VojtaStavik) [reply Tests/Virtual Time/VirtualTimers.swift:39]: We have it because both simple and repeating timers are represented by the same object in tests, while in the production, we handle them differently. `DispatchQueue.asyncAfter` for simple timers, and `DispatchSourceTimer` for repeating timers.

I'll change it so it has the api you want.

File: Tests/Client/Unit Tests/WebSocket/WebSocketTests.swift (hunk 3 above)
Comment (VojtaStavik) [reply Tests/Client/Unit Tests/WebSocket/WebSocketTests.swift:57]: I'm not super excited about this either. However, I ended up with the almost exact implementation having there in two functions. If you feel strongly about this, I'll change it. But it's a worse-of-two-evils kind of decision.

Comment (b-onc) [reply Tests/Client/Unit Tests/WebSocket/WebSocketTests.swift:57]: OK Let's keep it then 👍

File: Sources/Client/WebSocket/WebSocket.swift (hunk 5 above)
[the remark this answers is not in the document]
Comment (VojtaStavik) [reply Sources/Client/WebSocket/WebSocket.swift:109]: I'll remove `DispatchTimeInterval` completely and use just `TimeInterval`.

[the remark this answers is not in the document]
Comment (b-onc) [reply Sources/Client/WebSocket/WebSocket.swift:109]: Clarification: "Not something I'd push" => Agreed, let's move on :D Fine as i is 👍

[the remark this answers is not in the document]
Comment (VojtaStavik) [reply Sources/Client/WebSocket/WebSocket.swift:109]: Too late. I understood "I'd be ashamed to push code like that..." Anyway, we have better API now :)

[the remark this answers is not in the document]
Comment (b-onc) [reply Sources/Client/WebSocket/WebSocket.swift:109]: OK, all the better if you think it's better!

File: Tests/Virtual Time/VirtualTimers.swift [hunk 7]
```diff
@@ -12,46 +12,19 @@ struct VirtualTimeTimer: StreamChatClient.Timer {
...
-        case .nanoseconds(let value):
-            self = Double(value) / 1_000_000_000
-        case .never:
-            self = .greatestFiniteMagnitude
-        @unknown default:
-            fatalError()
-        }
-    }
-}
+extension VirtualTime.TimerControl: TimerControl { }
```
Comment (b-onc) [inline Tests/Virtual Time/VirtualTimers.swift:30]: Why don't we put it in `VirtualTime.swift` file, or directly to implementation?

File: Tests/Virtual Time/VirtualTime.swift [hunk 8]
```diff
@@ -113,5 +130,7 @@ extension VirtualTime {
             
             return nil
         }
+        
+
```
Comment (b-onc) [inline Tests/Virtual Time/VirtualTime.swift:134]: Unnecessary whitespace?

File: Tests/Virtual Time/VirtualTimers.swift (hunk 7 above)
Comment (VojtaStavik) [reply Tests/Virtual Time/VirtualTimers.swift:30]: `VirtualTime.swift` is a file independent to `StreamChatClient`. You can take it and include to any target you want. `VirtualTimers.swift` is where `VirtualTime` is used to implement internal protocols `StreamChatClient .Timer` and `StreamChatClient.TimerControl`. It's good to have them separate.
